hamilton/function_modifiers/macros.py

```diff
@@ -289,3 +303,515 @@ def __init__(self, model_cls, config_param: str, **extra_model_params):
+    problem, and is meant to represent a pipeline of chaining/redefinitions. This is similar (and can happily be
+    used in conjunction with) `pipe` in pandas. In Pyspark this is akin to the common operation of redefining a dataframe
+    with new columns. While it is generally reasonable to contain these constructs within a node's function,
+    you should consider `pipe` for any of the following reasons:
+
+    1.  You want the transformations to display as nodes in the DAG, with the possibility of storing or visualizing
+    the result
+    2. You want to pull in functions from an external repository, and build the DAG a little more procedurally
+    3. You want to use the same function multiple times, but with different parameters -- while `@does`/`@parameterize` can
+    do this, this presents an easier way to do this, especially in a chain.
+
+    To demonstrate the rules for chaining nodes, we'll be using the following example. This is
+    using primitives to demonstrate, but as hamilton is just functions of any python objects, this works perfectly with
+    dataframes, series, etc...
+
+
+    .. code-block:: python
+
+        def _add_one(x: int) -> int:
+            return x + 1
+
+        def _sum(x: int, y: int) -> int:
+            return x + y
+
+        def _multiply(x: int, y: int, z: int=10) -> int:
+            return x * y * z
+
+        @pipe(
+            apply(_add_one),
+            apply(_multiply, y=2),
+            apply(_sum, y=value(3)),
+            apply(_multiply, y=source("upstream_node_to_multiply")),
+        )
+        def final_result(upstream_int: int) -> int:
+            return upstream_int
+
+    .. code-block:: python
+
+        upstream_int = ... # result from upstream
+        upstream_node_to_multiply = ... # result from upstream
+        final_result(
+            _multiply(
+                _sum(
+                    _multiply(
+                        _add_one(upstream_int),
+                        y=2
+                    ),
+                    y=3
+                ),
+                y=upstream_node_to_multiply
+            )
+        )
+
+
+    Note that functions must have no position-only arguments (this is rare in python, but hamilton does not handle these).
+    This basically means that the functions must be defined similarly to `def fn(x, y, z=10)` and not `def fn(x, y, /, z=10)`.
+    In fact, all arguments must be named and "kwarg-friendly", meaning that the function can happily be called with `**kwargs`,
+    where kwargs are some set of resolved upstream values. So, no `*args` are allowed, and `**kwargs` (variable keyword-only) are not
+    permitted. Note that this is not a design limitation, rather an implementation detail -- if you feel like you need this, please
+    reach out.
+
+    Furthermore, the function should be typed, as a Hamilton function would be.
+
+    One has two ways to tune the shape/implementation of the subsequent nodes:
```

Comment:
should say why.

_rather than if else statements within code and functions taking these configuration parameters_ 
...